### PR TITLE
Development

### PR DESCRIPTION
**Pull request by cmceachin5888**
first pull development to main

### CODE REVIEW

File: db/models/billingAddress.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    } catch (error) {
+      throw new Error('Could not create billing address: ' + error.message);
+    }
+}
+
+async function addBillingAddressToUser(userId, billingAddressList) {
+  try {
+      const createdBillingAddresses = [];
+      for (const billingAddress of billingAddressList) {
+          const createdAddress = await createBillingAddress(
```
Comment (parisyee) [inline db/models/billingAddress.js:24]: named parameters would be better here.

**Review by parisyee — APPROVED** [review]
Great work so far! Just a few minor things but in general it looks really awesome. Can't wait for the frontend.

File: db/models/billingAddress.js (hunk 1 above)
Comment (cmceachin5888) [reply db/models/billingAddress.js]: I will review

File: db/models/products.js [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+      description , 
+      min_price, 
+      max_price, 
+      currency_code, 
+      amount, 
+      availability,
+      total_inventory     
+    } = updatedFields;
+
+    if (requestingUserRole !== 'admin') {
```
[the remark this answers is not in the document]
Comment (cmceachin5888) [reply db/models/products.js:138]: i was debating and determined that we could have DB verification and then also have router verification. i didnt think more would hurt. 

we do have questions about the best way to do admin verification and updating.

File: api/products.js [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  }
+
+  try {
+    const decodedToken = jwt.verify(token, process.env.JWT_ADMIN);
+    const userId = decodedToken?.id;
+
+    if (!userId) {
+      return res.status(401).json({
+        error: "Invalid token",
+        message: "You must be logged in as an admin to perform this action",
```
Comment (leafoflegend) [inline api/products.js:47]: Have you confirmed they are an `admin`? It looks like you've only confirmed they are a `user`, though I do infact see the `process.env.JWT_ADMIN`. I'll probably figure that out when I get lower to users.

File: api/reviews.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
+const express = require('express');
+const router = express.Router();
+const bodyParser = require('body-parser');
+
+router.use(bodyParser.json());
```
Comment (leafoflegend) [inline api/reviews.js:5]: Ditto previous.

File: api/reviews.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+const {
+  createReview,
+  getAllReviews,
+  getReviewsByProduct,
+} = require('../db/models/review');
+
+// POST /api/reviews
+router.post('/', async (req, res) => {
+  try {
+    const { productId, userId, rating, reviewText, reviewDate } = req.body;
```
Comment (leafoflegend) [inline api/reviews.js:16]: I could post a review pretending to be any user if I have their `userId` handy.

File: api/users.js [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+      const userId = decodedToken.id;
+      const user = await getUserById(userId);
+      if (!user) {
+        return res.status(401).json({
+          error: "User not found",
+          message: "You must be logged in to perform this action",
+          name: "UnauthorizedError",
+        });
+      }
+      res.json(user);
```
Comment (leafoflegend) [inline api/users.js:147]: Does this include their password in this information (I understand its encrypted, but still...).

File: api/users.js [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+                name: 'UserPasswordError'
+            });
+        }
+        const newUser = await createUser({ username, password });
+      const token = jwt.sign(
+        {
+          id: newUser.id,
+          username: newUser.username,
+        },
+        process.env.JWT_SECRET,
```
Comment (leafoflegend) [inline api/users.js:48]: So the only way to end up an admin is if you were originally manually encrypted with that secret?

File: app.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
+require("dotenv").config();
+const express = require('express');
+const app = express();
+const cors = require('cors');
+const morgan = require('morgan');
+
+// Setup your Middleware
+app.use(cors());
+app.use(morgan('dev'));
+app.use(express.json());
```
Comment (leafoflegend) [inline app.js:10]: These save you from having to use `bodyParser` at all.

Also, you could be some authentication middleware here that appends whether the user is authenticated or not.

File: db/.DS_Store
Comment (leafoflegend) [inline db/.DS_Store:1]: `.gitignore`

File: app.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
+require("dotenv").config();
```
Comment (leafoflegend) [inline app.js:1]: This should be the only place this is required.

File: server.js [hunk 10]
```diff
@@ -0,0 +1,14 @@
+const http = require("http")
+const chalk = require("chalk")
+const app = require("./app")
+
+const PORT = process.env["PORT"] || 3000
```
Comment (leafoflegend) [inline server.js:5]: You haven't used `dotenv` here, so this won't work. You can move all the entry point logic to this file.

File: server.js [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+const server = http.createServer(app)
+
+server.listen(PORT, () => {
+  console.log(
+    chalk.blueBright("Server is listening on PORT:"),
+    chalk.yellow(PORT),
+    chalk.blueBright("Get your Bike on!")
```
Comment (leafoflegend) [inline server.js:12]: ❤️ `chalk` ❤️

File: src/components/App.js [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    // invoke it immediately after its declaration, inside the useEffect callback
+    getAPIStatus();
+  }, []);
+
+  return (
+    <div className="app-container">
+      <h1>Hello, World!</h1>
+      <p>API Status: {APIHealth}</p>
+    </div>
+  );
```
Comment (leafoflegend) [inline src/components/App.js:30]: This is the whole frontend right now?

**Review by leafoflegend — CHANGES_REQUESTED** [review]
There are quite a lot of security vulnerabilities in here.

My list priority ordered of how you should approach fixes:

1. Security commentary.
2. DRY up the app.
3. Fix entry point issues - make sure things that only need to be called once aren't called various times across the app.
4. Add a linter.
5. Time to get to frontend!

Loved that you have tests setup! That is a **GREAT** thing!

**Pull request merged**